pkg/apiserver/rest/dualwriter_mode2.go

```diff
@@ -410,3 +416,200 @@ func enrichLegacyObject(originalObj, returnedObj runtime.Object) error {
+	metrics.init(reg)
+
+	log := klog.NewKlogr().WithName("DualWriterMode2Syncer")
+
+	everythingSynced := false
+	outOfSync := 0
+	syncSuccess := 0
+	syncErr := 0
+
+	maxInterval := dataSyncerInterval + 5*time.Minute
+
+	// LockExecuteAndRelease ensures that just a single Grafana server acquires a lock at a time
+	// The parameter 'maxInterval' is a timeout safeguard, if the LastExecution in the
+	// database is older than maxInterval, we will assume the lock as timeouted. The 'maxInterval' parameter should be so long
+	// that is impossible for 2 processes to run at the same time.
+	err := serverLockService.LockExecuteAndRelease(ctx, "dualwriter mode 2 sync", maxInterval, func(context.Context) {
+		log.Info("starting dualwriter mode 2 sync")
+		startSync := time.Now()
+
+		orgId := int64(1)
+
+		ctx = klog.NewContext(ctx, log)
+		ctx = identity.WithRequester(ctx, getSyncRequester(orgId))
+		ctx = request.WithNamespace(ctx, requestInfo.Namespace)
+		ctx = request.WithRequestInfo(ctx, requestInfo)
+
+		legacyList, err := getList(ctx, legacy, &metainternalversion.ListOptions{})
+		if err != nil {
+			log.Error(err, "unable to extract list from legacy storage")
+			return
+		}
+		log.Info("got items from legacy storage", "items", len(legacyList))
+
+		storageList, err := getList(ctx, storage, &metainternalversion.ListOptions{})
+		if err != nil {
+			log.Error(err, "unable to extract list from storage")
+			return
+		}
+		log.Info("got items from unified storage", "items", len(storageList))
+
+		itemsByName := map[string]syncItem{}
+		for _, obj := range legacyList {
+			accessor, err := utils.MetaAccessor(obj)
+			if err != nil {
+				log.Error(err, "error retrieving accessor data for object from legacy storage")
+				continue
+			}
+			name := accessor.GetName()
+
+			item, ok := itemsByName[name]
+			if !ok {
+				item = syncItem{}
+			}
+			item.name = name
+			item.objLegacy = obj
+			itemsByName[name] = item
+		}
+
+		for _, obj := range storageList {
+			accessor, err := utils.MetaAccessor(obj)
+			if err != nil {
+				log.Error(err, "error retrieving accessor data for object from storage")
+				continue
+			}
+			name := accessor.GetName()
+
+			item, ok := itemsByName[name]
+			if !ok {
+				item = syncItem{}
+			}
+			item.name = name
+			item.objStorage = obj
+			itemsByName[name] = item
+		}
+		log.Info("got list of items to be synced", "items", len(itemsByName))
+
+		for name, item := range itemsByName {
+			// upsert if:
+			// - existing in both legacy and storage, but objects are different, or
+			// - if it's missing from storage
+			if item.objLegacy != nil &&
+				((item.objStorage != nil && !Compare(item.objLegacy, item.objStorage)) || (item.objStorage == nil)) {
+				outOfSync++
+
+				accessor, err := utils.MetaAccessor(item.objLegacy)
+				if err != nil {
+					log.Error(err, "error retrieving accessor data for object from storage")
+					continue
+				}
+				accessor.SetResourceVersion("")
+				accessor.SetUID("")
```

Comment:
could move these into the `else` case below

Reply:
Agree with that. Done!